File: solutions/python/easy/p0136_single_number.py
"""LeetCode 136: Single Number.

Problem: https://leetcode.com/problems/single-number/
"""

from typing import List


class Solution:
    def singleNumber(self, nums: List[int]) -> int:
        """Use XOR because equal values cancel each other out.

        Time: O(n)
        Space: O(1)
        """
        result = 0
        for number in nums:
            result ^= number
        return result

File: solutions/java/easy/P0136_SingleNumber.java
/**
 * LeetCode 136: Single Number.
 *
 * Problem: https://leetcode.com/problems/single-number/
 *
 * Time: O(n)
 * Space: O(1)
 *
 * Note: Reviewed with assistance on 2026-06-09.
 */
public class P0136_SingleNumber {
    public int singleNumber(int[] nums) {
        int result = 0;

        for (int number : nums) {
            result ^= number;
        }

        return result;
    }

    public static void main(String[] args) {
        P0136_SingleNumber solution = new P0136_SingleNumber();

        System.out.println(solution.singleNumber(new int[] {2, 2, 1}));       // Expected: 1
        System.out.println(solution.singleNumber(new int[] {4, 1, 2, 1, 2})); // Expected: 4
        System.out.println(solution.singleNumber(new int[] {1}));             // Expected: 1
    }
}
